File: Observer/Observer.php
<?php
/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/25
 * Time: 14:34
 */

class Subject{
    private $state;
    private $observers;
    public function getState(){
        return $this->state;
    }

    public function setState($state){
        $this->state=$state;
        $this->notifyUsers();

    }

    public function attach($observer){
        $this->observers[]=$observer;
    }


    public function notifyUsers(){
        foreach($this->observers as $observer){
            $observer->update();
        }
    }
}


abstract  class Observer{
    protected $subject;
    public abstract function update();
}


class BinaryObserver extends Observer{
    public function __construct($subject)
    {
        $this->subject=$subject;
        $this->subject->attach($this);
    }

    public function update()
    {
        // TODO: Implement update() method.
        var_dump("Binary String: ".$this->subject->getState());
    }
}

class OctalObserver extends Observer{
    public function __construct($subject)
    {
        $this->subject=$subject;
        $this->subject->attach($this);
    }

    public function update()
    {
        // TODO: Implement update() method.
        var_dump("Octal String: ".$this->subject->getState());
    }
}

class HexaObserver extends Observer{
    public function __construct($subject)
    {
        $this->subject=$subject;
        $this->subject->attach($this);
    }

    public function update()
    {
        // TODO: Implement update() method.
        var_dump("Hexa String: ".$this->subject->getState());
    }
}

//demo

$subject=new Subject();

new HexaObserver($subject);
new OctalObserver($subject);
new BinaryObserver($subject);

var_dump("First state change: 15");
$subject->setState(15);

var_dump("Secend state change: 10");
$subject->setState(10);

File: Observer/Observer.go
package main

import (
	"container/list"
	"fmt"
)

type Observer interface {
	update()
}

type Subject struct {
	state int
	observers *list.List
}

func(s *Subject)getState() int{
	return s.state
}

func(s *Subject)setState(state int){
	s.state=state
	s.notifyObservers()
}

func(s *Subject)attach(observer Observer){
	s.observers.PushBack(observer)
}

func(s *Subject)notifyObservers(){
	for o:=s.observers.Front();o!=nil;o=o.Next(){
		observer:=o.Value.(Observer)
		observer.update()
	}
}


type BinaryObserver struct {
	subject *Subject
}

func (b *BinaryObserver)init(subject *Subject){
	b.subject=subject
	b.subject.attach(b)
}

func(b *BinaryObserver)update(){
	fmt.Println("this is binary string",b.subject.getState())
}

type OctalObserver struct {
	subject *Subject
}

func (b *OctalObserver)init(subject *Subject){
	b.subject=subject
	b.subject.attach(b)
}

func(b *OctalObserver)update(){
	fmt.Println("this is octal string",b.subject.getState())
}


type HexaObserver struct {
	subject *Subject
}

func (b *HexaObserver)init(subject *Subject){
	b.subject=subject
	b.subject.attach(b)
}

func(b *HexaObserver)update(){
	fmt.Println("this is hexa string",b.subject.getState())
}


func main(){
	subject:=&Subject{observers:list.New()};
	o1:=new(BinaryObserver)
	o1.init(subject)
	o2:=new(OctalObserver)
	o2.init(subject)
	o3:=new(HexaObserver)
	o3.init(subject)

	fmt.Println("first state changes:15")
	subject.setState(15)

	fmt.Println("secend state changes:10")
	subject.setState(10)
}